### PR TITLE
feat: add file type validation and mime type extraction

### PR DESCRIPTION
**Pull request by jasonlai1218**
# TL;DR
add file type validation and mime type extraction

## Type
 - [x] Bug Fix
 - [ ] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [ ] Smoke tested
 - [x] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
- Import the `magic` module in `flytekit/types/file/file.py`
- Add a method `get_mime_type_from_python_type` in `FlyteFilePathTransformer` class in `flytekit/types/file/file.py`
- Add a validation for file type in `FlyteFilePathTransformer.to_literal` method in `flytekit/types/file/file.py`
- Modify `setup.py` to include `python-magic>=0.4.27` as a dependency
- Add a test case `test_real_file_type_in_workflow` in `test_flyte_file.py`

### (Scenario: actual file type does not match the expected file type)
Screenshot 2023-10-15 at 11 34 13 PM

### (Scenario: actual file type matches the expected file type)
Screenshot 2023-10-15 at 11 35 05 PM

### (Scenario: actual pdf file type does match the expected pdf file type)
Screenshot 2023-10-15 at 11 33 01 PM

---

### Remote Test
`docker build . -f Dockerfile.dev -t localhost:30000/flytekit:dev1 --build-arg PYTHON_VERSION=3.9 
`
`docker push localhost:30000/flytekit:dev1 `
` pyflyte run --remote --image localhost:30000/flytekit:dev1 jpg_png_mismatch.py wf `

### (Scenario: actual file type does not match the expected file type)
Screenshot 2023-10-17 at 9 50 37 AM

## Tracking Issue
Fixes https://github.com/flyteorg/flyte/issues/3834

## Follow-up issue
_NA_

### CODE REVIEW

**Comment by parag477** [comment]
I want to work on this issue.

File: tests/flytekit/unit/core/test_flyte_file.py [hunk 1]
```diff
@@ -52,6 +52,79 @@ def my_wf() -> FlyteFile[typing.TypeVar("txt")]:
...
+    source_path = "/tmp/flytekit_test.png"
+    source_file_mime_type = "image/png"
+    user_defined_format = "jpeg"
+
+    FlyteFilePathTransformer.get_format = MagicMock(return_value=user_defined_format)
+    with patch("magic.from_file", return_value=source_file_mime_type):
+        with pytest.raises(ValueError, match=f"Incorrect file type, expected image/jpeg, got {source_file_mime_type}"):
+            transformer.validate_file_type(user_defined_format, source_path)
+
+
```
Comment (eapolinario) [inline tests/flytekit/unit/core/test_flyte_file.py:127]: Can you add test where `FlyteFile` is `Annotated`?

File: flytekit/types/file/file.py [hunk 2]
```diff
@@ -6,6 +6,7 @@
 from contextlib import contextmanager
 from dataclasses import dataclass, field
 
+import magic
```
[the remark this answers is not in the document]
Comment (jasonlai1218) [reply flytekit/types/file/file.py:9]: Thank you for providing the example. : )

File: flytekit/types/file/file.py [hunk 3]
```diff
@@ -320,6 +321,31 @@ def assert_type(
...
+            "ipynb": "application/x-ipynb+json",
+            "svg": "image/svg+xml",
+            "csv": "text/csv",
+            "onnx": "application/octet-stream",
+            "tfrecord": "application/octet-stream",
+            "txt": "text/plain",
+        }
+        return extension_to_mime_type[extension]
+
+    def validate_file_type(self, python_type, source_path):
```
[the remark this answers is not in the document]
Comment (jasonlai1218) [reply flytekit/types/file/file.py:342]: sure, thanks for the reminder

File: flytekit/types/file/file.py [hunk 4]
```diff
@@ -320,6 +321,31 @@ def assert_type(
...
+        if FlyteFilePathTransformer.get_format(python_type):
```
[the remark this answers is not in the document]
Comment (jasonlai1218) [reply flytekit/types/file/file.py:343]: That's right.

File: tests/flytekit/unit/core/test_flyte_file.py [hunk 5]
```diff
@@ -52,6 +52,79 @@ def my_wf() -> FlyteFile[typing.TypeVar("txt")]:
         assert fh.read() == "Hello World\n"
 
 
+def test_matching_file_types_in_workflow():
```
[the remark this answers is not in the document]
Comment (jasonlai1218) [reply tests/flytekit/unit/core/test_flyte_file.py:55]: I've added [here](https://github.com/flyteorg/flytekit/blob/b64f06d7f28494118afde508e0d30f575bb82080/tests/flytekit/unit/core/test_flyte_file.py#L75-L91)

File: flytekit/types/file/file.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (jasonlai1218) [reply flytekit/types/file/file.py:350]: I've added [here].(https://github.com/flyteorg/flytekit/blob/b64f06d7f28494118afde508e0d30f575bb82080/flytekit/types/file/file.py#L342-L348)
I think maybe we can use try-except in a simple way to achieve the desired purpose. Is this OK?

File: tests/flytekit/unit/core/test_flyte_file.py (hunk 1 above)
Comment (jasonlai1218) [reply tests/flytekit/unit/core/test_flyte_file.py:146]: Sorry, I don't understand what you mean. Can you give me some more hints or suggestions?

Comment (jasonlai1218) [reply tests/flytekit/unit/core/test_flyte_file.py:180]: I've added here.
https://github.com/flyteorg/flytekit/blob/35ddf821a2985c26faf6c5116e387a44f091719d/tests/flytekit/unit/core/test_flyte_file.py#L159-L178

**Comment by jasonlai1218** [comment]
@eapolinario Please help review PR, thank you

File: flytekit/types/file/file.py [hunk 6]
```diff
@@ -324,6 +325,44 @@ def assert_type(
...
+        }
+        return extension_to_mime_type[extension]
+
+    def validate_file_type(self, python_type: typing.Type[FlyteFile], source_path: typing.Union[str, os.PathLike]):
+        try:
+            # isolate the exception to the libmagic import
+            import magic
+
+        except ImportError as e:
+            logger.warning(f"Libmagic is not installed. Error message: {e}")
```
Comment (eapolinario) [inline flytekit/types/file/file.py:352]: this should not be a warning. Let's use `debug`.

File: flytekit/types/file/file.py [hunk 7]
```diff
@@ -324,6 +325,44 @@ def assert_type(
...
+
+        if FlyteFilePathTransformer.get_format(python_type):
+            try:
+                real_type = magic.from_file(source_path, mime=True)
+                expected_type = self.get_mime_type_from_python_type(FlyteFilePathTransformer.get_format(python_type))
+                if real_type != expected_type:
+                    raise ValueError(f"Incorrect file type, expected {expected_type}, got {real_type}")
+            except FileNotFoundError:
+                # This scenario occurs during unit testing.
+                # If the actual file does not exist, no comparison of file types is needed.
```
Comment (eapolinario) [inline flytekit/types/file/file.py:363]: what test caused this?

File: flytekit/types/file/file.py [hunk 8]
```diff
@@ -324,6 +325,44 @@ def assert_type(
...
+            "hdf5": "text/plain",
+            "joblib": "application/octet-stream",
+            "python_pickle": "application/octet-stream",
+            "ipynb": "application/json",
+            "onnx": "application/json",
+            "tfrecord": "application/octet-stream",
+        }
+        return extension_to_mime_type[extension]
+
+    def validate_file_type(self, python_type: typing.Type[FlyteFile], source_path: typing.Union[str, os.PathLike]):
```
[the remark this answers is not in the document]
Comment (jasonlai1218) [reply flytekit/types/file/file.py]: Updated it

File: setup.py [hunk 9]
```diff
@@ -75,6 +75,7 @@
         "rich",
         "rich_click",
         "jsonpickle",
+        "python-magic",
```
[the remark this answers is not in the document]
Comment (jasonlai1218) [reply setup.py]: got it, thanks for the reminder

File: flytekit/types/file/file.py (hunk 7 above)
Comment (jasonlai1218) [reply flytekit/types/file/file.py:374]: Because in the [test_flyte_file_in_dataclassjsonmixin unit test](https://github.com/flyteorg/flytekit/blob/b176572a0c97cefd64bbe0c28a6401839be3f134/tests/flytekit/unit/core/test_type_engine.py#L1045-L1074), the [validate_file_type](https://github.com/flyteorg/flytekit/blob/b176572a0c97cefd64bbe0c28a6401839be3f134/flytekit/types/file/file.py#L346-L375) method will be passed, in which [magic.from_file](https://github.com/flyteorg/flytekit/blob/b176572a0c97cefd64bbe0c28a6401839be3f134/flytekit/types/file/file.py#L368) will check the file. If the file does not exist, a **FileNotFoundException** will be thrown.

Comment (jasonlai1218) [reply flytekit/types/file/file.py:374]: as shown in the picture
Screenshot 2023-10-31 at 2 51 22 PM

Comment (jasonlai1218) [reply flytekit/types/file/file.py:374]: That’s why I used **`try-except`** to handle it. Is it appropriate to handle it this way? Is there any other better way?

File: tests/flytekit/unit/core/test_flyte_file.py [hunk 10]
```diff
@@ -34,6 +34,27 @@ def local_dummy_file():
...
+
+
+@pytest.fixture
+def can_import_magic():
+    try:
+        import magic  # noqa: F401
+
+        return True
+    except ImportError:
+        return False
```
[the remark this answers is not in the document]
Comment (jasonlai1218) [reply tests/flytekit/unit/core/test_flyte_file.py]: Thanks for the great advice. I learned new knowledge : )

[the remark this answers is not in the document]
Comment (jasonlai1218) [reply tests/flytekit/unit/core/test_flyte_file.py]: I updated it.
like this
https://github.com/flyteorg/flytekit/blob/b176572a0c97cefd64bbe0c28a6401839be3f134/tests/flytekit/unit/core/test_flyte_file.py#L98-L111

**Comment by samhita-alla** [comment]
@jasonlai1218, can you resolve the merge conflict?

File: flytekit/types/file/file.py (hunk 7 above)
Comment (jasonlai1218) [reply flytekit/types/file/file.py:374]: After thinking for a while, I think it is not right to write try-except in order to let the unit test pass. Maybe we should adjust the unit test.
I think the simple thing to do is to skip this test if python-magic is installed, what do you think?

Comment (jasonlai1218) [reply flytekit/types/file/file.py]: I removed the try-except and adjusted the unit tests.
https://github.com/flyteorg/flytekit/blob/7ac4c8f840fb587d9a1df3ed0aaaf40553a8e485/tests/flytekit/unit/core/test_type_engine.py#L1047-L1080

Comment (eapolinario) [reply flytekit/types/file/file.py]: One of the usecases for FlyteFile is to point to remote files, in other words, you might have access to a remote file (e.g. in s3) that you want to pass to a Flyte workflow. 

More concretely, we should only validate FlyteFiles for which their `path` is considered local (via a call to [FileAccessProvider.is_remote](https://github.com/flyteorg/flytekit/blob/master/flytekit/core/data_persistence.py#L184-L191)).

Comment (jasonlai1218) [reply flytekit/types/file/file.py]: OK, I updated it.
like this
https://github.com/flyteorg/flytekit/blob/c1e2774cca85f84f9d7a0a80d140c5591ae0bd54/tests/flytekit/unit/core/test_type_engine.py#L1010-L1019

Comment (jasonlai1218) [reply flytekit/types/file/file.py]: https://github.com/flyteorg/flytekit/blob/06bd4c4d2cdfdd46119342c58e51e69cc8687360/flytekit/types/file/file.py#L358-L359

File: tests/flytekit/unit/extras/tasks/testdata/test.csv [hunk 11]
```diff
@@ -0,0 +1,4 @@
+SN,Name,Contribution
```
Comment (pingsutw) [inline tests/flytekit/unit/extras/tasks/testdata/test.csv:1]: should we remove this file?

File: tests/flytekit/unit/core/test_type_hints.py [hunk 12]
```diff
@@ -390,7 +390,7 @@ def test_flyte_file_in_dataclass():
     @dataclass
     class InnerFileStruct(DataClassJsonMixin):
         a: FlyteFile
-        b: PNGImageFile
+        b: FlyteFile
```
Comment (pingsutw) [inline tests/flytekit/unit/core/test_type_hints.py:393]: why do we change the type to `FlyteFile` from `PNGImageFile` here?

File: tests/flytekit/unit/extras/tasks/testdata/test.csv (hunk 11 above)
Comment (jasonlai1218) [reply tests/flytekit/unit/extras/tasks/testdata/test.csv:1]: we need this file now.
currently this test.csv file is used in [test_shell.py](https://github.com/jasonlai1218/flytekit/blob/add-validation-flyteformat-type/tests/flytekit/unit/extras/tasks/test_shell.py#L17-L19)
https://github.com/jasonlai1218/flytekit/blob/add-validation-flyteformat-type/tests/flytekit/unit/extras/tasks/test_shell.py#L17-L19

The reason why I adjusted this file is that although its file extension is .csv, it is not actually a csv file format, so I adjusted it to the csv format type.

File: tests/flytekit/unit/core/test_type_hints.py (hunk 12 above)
Comment (jasonlai1218) [reply tests/flytekit/unit/core/test_type_hints.py:393]: Because it is declared as PNGImageFIle, it will be checked whether the declared file type is consistent with the actual file format. However, in the current unit test, there is no actual file, which will cause a FileNotFoundError.
like this
Screenshot 2023-11-07 at 9 03 06 AM

https://github.com/jasonlai1218/flytekit/blob/add-validation-flyteformat-type/tests/flytekit/unit/core/test_type_hints.py#L426-L430

Comment (jasonlai1218) [reply tests/flytekit/unit/core/test_type_hints.py:393]: Screenshot 2023-11-07 at 10 03 44 AM

File: tests/flytekit/unit/extras/tasks/testdata/test.csv (hunk 11 above)
Comment (pingsutw) [reply tests/flytekit/unit/extras/tasks/testdata/test.csv:1]: i see, thanks

File: flytekit/types/file/file.py [hunk 13]
```diff
@@ -324,6 +325,60 @@ def assert_type(
     def get_literal_type(self, t: typing.Union[typing.Type[FlyteFile], os.PathLike]) -> LiteralType:
         return LiteralType(blob=self._blob_type(format=FlyteFilePathTransformer.get_format(t)))
 
+    def get_mime_type_from_python_type(self, extension: str) -> str:
```
Comment (pingsutw) [inline flytekit/types/file/file.py:328]: LGTM, one nit: could we rename it to `get_mime_type_from_extension`?

File: tests/flytekit/unit/core/test_flyte_file.py [hunk 14]
```diff
@@ -52,6 +71,116 @@ def my_wf() -> FlyteFile[typing.TypeVar("txt")]:
...
+    def my_wf(path: FlyteFile[typing.TypeVar("txt")]) -> FlyteFile[typing.TypeVar("jpeg")]:
+        f = t1(path=path)
+        return f
+
+    with pytest.raises(TypeError) as excinfo:
+        my_wf(path=local_dummy_txt_file)
+    assert "Incorrect file type, expected image/jpeg, got text/plain" in str(excinfo.value)
+
+
+def test_get_mime_type_from_python_type_success():
```
Comment (pingsutw) [inline tests/flytekit/unit/core/test_flyte_file.py:121]: ditto
```suggestion
def test_get_mime_type_from_extension_success():
```

File: tests/flytekit/unit/core/test_flyte_file.py [hunk 15]
```diff
@@ -52,6 +71,116 @@ def my_wf() -> FlyteFile[typing.TypeVar("txt")]:
...
+    assert transformer.get_mime_type_from_python_type("python_pickle") == "application/octet-stream"
+    assert transformer.get_mime_type_from_python_type("ipynb") == "application/json"
+    assert transformer.get_mime_type_from_python_type("svg") == "image/svg+xml"
+    assert transformer.get_mime_type_from_python_type("csv") == "text/csv"
+    assert transformer.get_mime_type_from_python_type("onnx") == "application/json"
+    assert transformer.get_mime_type_from_python_type("tfrecord") == "application/octet-stream"
+    assert transformer.get_mime_type_from_python_type("txt") == "text/plain"
+
+
+def test_get_mime_type_from_python_type_failure():
```
Comment (pingsutw) [inline tests/flytekit/unit/core/test_flyte_file.py:138]:
```suggestion
def test_get_mime_type_from_extension_failure():
```

File: flytekit/types/file/file.py [hunk 16]
```diff
@@ -324,6 +325,60 @@ def assert_type(
...
+        extension_to_mime_type = {
+            "html": mimetypes.types_map[".html"],
```
Comment (pingsutw) [inline flytekit/types/file/file.py:330]: could we append `mimetypes.types_map` to the `extension_to_mime_type`?

something like
```python
for extension, mimetype in range(mimetypes.types_map.items()):
    extension_to_mime_type[k.split(".")[1]] = mimetype
```

Comment (pingsutw) [reply flytekit/types/file/file.py:330]: btw, some tests are failing

File: flytekit/types/file/file.py (hunk 13 above)
Comment (jasonlai1218) [reply flytekit/types/file/file.py]: updated

File: tests/flytekit/unit/core/test_flyte_file.py (hunk 14 above)
Comment (jasonlai1218) [reply tests/flytekit/unit/core/test_flyte_file.py]: updated

File: tests/flytekit/unit/core/test_flyte_file.py (hunk 15 above)
Comment (jasonlai1218) [reply tests/flytekit/unit/core/test_flyte_file.py]: updated

File: flytekit/types/file/file.py (hunk 13 above)
Comment (jasonlai1218) [reply flytekit/types/file/file.py]: https://github.com/flyteorg/flytekit/blob/ec732bda3083e8ae23fd42476fa6c1be4f87ebff/flytekit/types/file/file.py#L328

File: flytekit/types/file/file.py (hunk 16 above)
Comment (jasonlai1218) [reply flytekit/types/file/file.py]: updated and all tests passed

File: tests/flytekit/unit/core/test_flyte_file.py (hunk 14 above)
Comment (jasonlai1218) [reply tests/flytekit/unit/core/test_flyte_file.py]: https://github.com/flyteorg/flytekit/blob/ec732bda3083e8ae23fd42476fa6c1be4f87ebff/tests/flytekit/unit/core/test_flyte_file.py#L121

File: flytekit/types/file/file.py (hunk 16 above)
Comment (jasonlai1218) [reply flytekit/types/file/file.py]: Screenshot 2023-11-07 at 11 43 50 AM

**Comment by eapolinario** [comment]
Something weird is happening to the windows tests. They are hanging, [running](https://github.com/flyteorg/flytekit/actions/runs/6793164167/job/18468575359?pr=1893) for multiple hours until they are canceled.

Can you take a look, @jasonlai1218 ?

**Comment by jasonlai1218** [comment]
@eapolinario 
According to the [doc](https://pypi.org/project/python-magic/), the lib that Windows OS needs to install is `python-magic-bin`, not `libmagic1`. I guess this is the cause of the build error.

Screenshot 2023-11-08 at 11 49 56 PM

so I adjusted it like this
https://github.com/flyteorg/flytekit/blob/aea3a47157c6bdaca9d3089e741f8fa4b36bc729/dev-requirements.in#L30-L33

**Comment by jasonlai1218** [comment]
> Something weird is happening to the windows tests. They are hanging, [running](https://github.com/flyteorg/flytekit/actions/runs/6793164167/job/18468575359?pr=1893) for multiple hours until they are canceled.
> 
> Can you take a look, @jasonlai1218 ?

**Comment by eapolinario** [comment]
@jasonlai1218 , how about we disable the feature on windows? Leave a TODO explaining the limitation and create a github issue.

**Comment by jasonlai1218** [comment]
> @jasonlai1218 , how about we disable the feature on windows? Leave a TODO explaining the limitation and create a github issue.

OK, I updated it.
https://github.com/flyteorg/flytekit/blob/63e634a7da34e9b45a7191b0776d0da11ce155ce/dev-requirements.in#L30-L34

and I create a github issue
https://github.com/flyteorg/flyte/issues/4391

**Review by eapolinario — APPROVED** [review]
